File: apps/python/face_verification.py
import pathlib
import os

from oneML import faceAPI as api
import numpy as np

os.environ['ONEML_CPP_MIN_LOG_LEVEL'] = "ERROR"
assets_path = os.path.join(
    str(pathlib.Path(__file__).resolve().parent.parent.parent.absolute()),
    "assets/images",
)


def main():
    manager = api.LicenseManager()
    manager.activate_trial()
    detector = api.FaceDetector(manager)
    embedder = api.FaceEmbedder(manager)
    face_id = api.FaceId(embedder, manager)
    utils = api.Utils(manager)

    img1 = utils.read_image_cv(
        os.path.join(
            assets_path, "register-set/Colin_Powell/colin_powell_0074.jpg"
        )
    )

    img2 = utils.read_image_cv(
        os.path.join(
            assets_path, "evaluate-set/Colin_Powell/colin_powell_0097.jpg"
        )
    )

    img3 = utils.read_image_cv(
        os.path.join(
            assets_path, "register-set/George_Robertson/george_robertson_0000.jpg"
        )
    )

    img4 = utils.read_image_cv(
        os.path.join(
            assets_path, "evaluate-set/George_Robertson/george_robertson_0009.jpg"
        )
    )

    register_images = np.stack((img1, img3), axis=0)
    eval_images = np.stack((img2, img4), axis=0)

    register_results = detector.detect_batch(register_images)
    eval_results = detector.detect_batch(eval_images)

    align_img1 = utils.crop_align_face_landmark(
        img1, register_results[0].get_landmarks()[0]
    )
    align_img3 = utils.crop_align_face_landmark(
        img3, register_results[1].get_landmarks()[0]
    )
    align_img2 = utils.crop_align_face_landmark(
        img2, eval_results[0].get_landmarks()[0]
    )
    align_img4 = utils.crop_align_face_landmark(
        img4, eval_results[1].get_landmarks()[0]
    )

    face_id.register_id_image("colin_powell", align_img1)
    face_id.register_id_image("george_robertson", align_img3)

    results = face_id.predict_batch(np.stack((align_img2, align_img4), axis=0))

    # Run with registration & predict style to print scores for validating with other language applications
    # since is_the_same_person_img API doesn't provide any score.
    print(
        "First person nearest node distance:",
        "{:.8f}".format(results[0].get_nearest_node_distance()),
    )
    print(
        "Second person nearest node distance:",
        "{:.8f}".format(results[1].get_nearest_node_distance()),
    )
    print("First person:", results[0].get_id())
    print("Second person:", results[1].get_id())

    # Also test the actual face verification API (is_the_same_person_img)
    is_same = face_id.is_the_same_person_img(img1, img2)
    print("Is the same person (colin_powell): " + str(int(is_same)))

    is_same = face_id.is_the_same_person_img(img3, img4)
    print("Is the same person (george_robertson): " + str(int(is_same)))

    report = detector.get_usage()
    report.to_log()

    report = face_id.get_usage()
    report.to_log()


if __name__ == "__main__":
    main()

File: apps/golang/face_verification.go
package main

import (
	"faceapps/utility"
	"fmt"
	"oneml/face"
	"os"
	"path"
)

func main() {
	currentPath, _ := os.Getwd()
	testAssetPath := path.Join(currentPath, "../../assets/images")

	// Initialize oneML modules
	licenseManager := face.NewLicenseManager()
	licenseManager.ActivateTrial()

	faceDetector := face.NewFaceDetector(licenseManager)
	faceEmbedder := face.NewFaceEmbedder(licenseManager)
	faceId := face.NewFaceId(faceEmbedder, licenseManager)
	utils := face.NewUtils(licenseManager)

	// Load images
	img1 := utils.ReadImageCV(path.Join(testAssetPath, "register-set/Colin_Powell/colin_powell_0074.jpg"))
	img2 := utils.ReadImageCV(path.Join(testAssetPath, "evaluate-set/Colin_Powell/colin_powell_0097.jpg"))
	img3 := utils.ReadImageCV(path.Join(testAssetPath, "register-set/George_Robertson/george_robertson_0000.jpg"))
	img4 := utils.ReadImageCV(path.Join(testAssetPath, "evaluate-set/George_Robertson/george_robertson_0009.jpg"))

	registerImages := face.NewMultiImage()
	registerImages.Add(img1)
	registerImages.Add(img3)

	evalImages := face.NewMultiImage()
	evalImages.Add(img2)
	evalImages.Add(img4)

	registerResults := faceDetector.DetectBatch(registerImages)
	evalResults := faceDetector.DetectBatch(evalImages)

	alignImg1 := utils.CropAlignFaceLandmark(img1, registerResults.Get(0).GetLandmarks().Get(0))
	alignImg3 := utils.CropAlignFaceLandmark(img3, registerResults.Get(1).GetLandmarks().Get(0))
	alignImg2 := utils.CropAlignFaceLandmark(img2, evalResults.Get(0).GetLandmarks().Get(0))
	alignImg4 := utils.CropAlignFaceLandmark(img4, evalResults.Get(1).GetLandmarks().Get(0))

	faceId.RegisterIdImage("colin_powell", alignImg1)
	faceId.RegisterIdImage("george_robertson", alignImg3)

	predictImages := face.NewMultiImage()
	predictImages.Add(alignImg2)
	predictImages.Add(alignImg4)

	results := faceId.PredictBatch(predictImages)

	// Run with registration & predict style to print scores for validating with other language applications
	// since isTheSamePerson API doesn't provide any score.
	fmt.Printf("First person nearest node distance: %.8f\n", results.Get(0).GetNearestNodeDistance())
	fmt.Printf("Second person nearest node distance: %.8f\n", results.Get(1).GetNearestNodeDistance())
	fmt.Println("First person:", results.Get(0).GetId())
	fmt.Println("Second person:", results.Get(1).GetId())

	// Also test the actual face verification API (IsTheSamePerson)
	isSame := faceId.IsTheSamePersonImg(img1, img2)
	fmt.Println("Is the same person (colin_powell):", utility.BoolToInt(isSame))

	isSame = faceId.IsTheSamePersonImg(img3, img4)
	fmt.Println("Is the same person (george_robertson):", utility.BoolToInt(isSame))

	report := faceDetector.GetUsage()
	report.ToLog()

	report = faceId.GetUsage()
	report.ToLog()
}
